include/secp256k1_schnorr_adaptor.h

```diff
@@ -0,0 +1,138 @@
+    const unsigned char *key32,
+    const unsigned char *t33,
+    const unsigned char *xonly_pk32,
+    const unsigned char *algo,
+    size_t algolen,
+    void *data
+);
+
+/** A modified BIP-340 nonce generation function.
+ *
+ *  If a data pointer is passed, it is assumed to be a pointer to 32 bytes of
+ *  auxiliary random data as defined in BIP-340. If the data pointer is NULL,
+ *  the nonce derivation procedure follows BIP-340 by setting the auxiliary
+ *  random data to zero. The algo argument must be non-NULL, otherwise the
+ *  function will fail and return 0. The hash will be tagged with algo.
+ *  Therefore, to create BIP-340 compliant signatures, algo must be set to
+ *  "BIP0340/nonce" and algolen to 13.
+ */
+SECP256K1_API const secp256k1_adaptor_nonce_function_hardened secp256k1_adaptor_nonce_function_bip340;
+
+/** Create a Schnorr adaptor signature.
+ *
+ *  This function only signs 32-byte messages. If you have messages of a
+ *  different size (or the same size but without a context-specific tag
+ *  prefix), it is recommended to create a 32-byte message hash with
+ *  secp256k1_tagged_sha256 and then sign the hash. Tagged hashing allows
+ *  providing an context-specific tag for domain separation. This prevents
+ *  signatures from being valid in multiple contexts by accident.
+ *
+ *  Returns 1 on success, 0 on failure.
+ *  Args:    ctx: pointer to a context object (not secp256k1_context_static).
+ *  Out:   sig65: pointer to a 65-byte array to store the serialized signature.
+ *  In:    msg32: the 32-byte message being signed.
+ *       keypair: pointer to an initialized keypair.
+ *           t33: pointer to a 33-byte compressed adaptor point.
+ *    aux_rand32: 32 bytes of fresh randomness. While recommended to provide
+ *                this, it is only supplemental to security and can be NULL. A
+ *                NULL argument is treated the same as an all-zero one. See
+ *                BIP-340 "Default Signing" for a full explanation of this
+ *                argument and for guidance if randomness is expensive.
+ */
+SECP256K1_API int secp256k1_schnorr_adaptor_presign(
+    const secp256k1_context *ctx,
+    unsigned char *sig65,
+    const unsigned char *msg32,
+    const secp256k1_keypair *keypair,
+    const unsigned char *t33,
+    const unsigned char *aux_rand32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3) SECP256K1_ARG_NONNULL(4) SECP256K1_ARG_NONNULL(5);
+
+/** Extract an adaptor point T from the signature.
+ *
+ *  Returns 1 on success, 0 on failure.
+ *  Args:    ctx: pointer to a context object.
+ *  Out:      t33: pointer to a 33-byte array to store the compressed adaptor point.
+ *  In:     sig65: pointer to a 65-byte serialized signature.
+ *           msg32: the 32-byte message being signed.
+ *           pubkey: pointer to an x-only public key to verify with (cannot be NULL)
+ */
+SECP256K1_API int secp256k1_schnorr_adaptor_extract_t(
```

Comment:
I think we should rename this because `t` is confusing. Maybe `secp256k1_schnorr_adaptor_preextract`?

Reply:
Similarly after https://github.com/BlockstreamResearch/secp256k1-zkp/pull/268#discussion_r1377766484, `secp256k1_schnorr_adaptor_extract` or `secp256k1_schnorr_adaptor_extract_adaptor`?

Reply:
I think for simplicity, we can name extract_t as extract and extract_adaptor as extract_sec, so it matches the variable names for adaptor and sec adaptor